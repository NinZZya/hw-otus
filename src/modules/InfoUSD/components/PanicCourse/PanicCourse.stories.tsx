import React from "react";
// also exported from "@storybook/react" if you can deal with breaking changes in 6.1
import { Meta } from "@storybook/react/types-6-0";

import { PanicCourse, PanicCourseProp } from "./PanicCourse";

export default {
  title: "Modules/InfoUSD/Components/PanicCourse",
  component: PanicCourse,
} as Meta;

export const DontPanic: React.FC<PanicCourseProp> = () => (
  <PanicCourse value={1} indicatorColor="#ff0000" />
);

export const WithPanic: React.FC<PanicCourseProp> = () => (
  <PanicCourse value={100} indicatorColor="#ff0000" />
);
