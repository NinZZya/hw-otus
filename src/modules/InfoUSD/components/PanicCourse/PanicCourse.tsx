import React from "react";
import { Label, Card, Indicator } from "../../../../componets";
import { getRandomColor } from "../../../../utils/utils";

export interface PanicCourseProp {
  value: number;
  indicatorColor?: string;
}

export interface PanicCourseState {
  count: number;
}

const PANIC_RANGE = 1;
const MAX_COUNT = 3;

export class PanicCourse extends React.Component<
  PanicCourseProp,
  PanicCourseState
> {
  constructor(props: PanicCourseProp) {
    super(props);

    this.state = {
      count: 0,
    };
  }

  componentDidUpdate(): void {
    if (this.state.count === MAX_COUNT) {
      throw new Error("Stop! You are very afraid");
    } else {
      this.setState({
        count: this.state.count + 1,
      });
    }
  }

  shouldComponentUpdate(nextProps: PanicCourseProp): boolean {
    return nextProps.value - this.props.value > PANIC_RANGE;
  }

  render(): React.ReactNode {
    const { count } = this.state;
    const { indicatorColor } = this.props;
    const color = indicatorColor ? indicatorColor : getRandomColor();

    return (
      <Card>
        <Indicator type={"card"} color={color} />
        <span>{`If the course is grov than $${PANIC_RANGE}, you have panic`}</span>
        <span>{`If the panic counter is ${MAX_COUNT}, there will be an error`}</span>
        <Label value={`Panic count: ${count}`} />
      </Card>
    );
  }
}
