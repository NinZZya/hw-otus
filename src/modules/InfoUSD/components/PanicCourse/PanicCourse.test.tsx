import React from "react";
import renderer from "react-test-renderer";
import { PanicCourse } from "./PanicCourse";

it("MaxCourse render coorectly", () => {
  const tree = renderer
    .create(<PanicCourse value={0} indicatorColor={"#000"} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
