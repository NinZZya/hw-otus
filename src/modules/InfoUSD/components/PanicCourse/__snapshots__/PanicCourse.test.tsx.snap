// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxCourse render coorectly 1`] = `
<div
  className="css-1y7yulv"
>
  <span
    className="css-15rxe6g"
    color="#000"
  >
    â€¢
  </span>
  <span>
    If the course is grov than $1, you have panic
  </span>
  <span>
    If the panic counter is 3, there will be an error
  </span>
  <span
    className="css-l3kone"
  >
    Panic count: 0
  </span>
</div>
`;
